lib/sidekiq/worker.rb

```diff
@@ -351,12 +347,31 @@ def sidekiq_options(opts = {})
         super
       end
 
-      def client_push(item) # :nodoc:
+      def client_push(stringified_item) # :nodoc:
         pool = Thread.current[:sidekiq_via_pool] || get_sidekiq_options["pool"] || Sidekiq.redis_pool
-        stringified_item = item.transform_keys(&:to_s)
-
         Sidekiq::Client.new(pool).push(stringified_item)
       end
+
+      def serialize(opts = {}, args = [])
+        new(opts, *args).serialize
+      end
     end
+
+    # Creates a new job instance. Takes the arguments that will be
+    # passed to the perform method.
+    def initialize(opts_or_args = {}, *args)
+      @opts = opts_or_args.is_a?(Hash) ? opts_or_args : {}
+      @args = opts_or_args.is_a?(Hash) ? args : Array[opts_or_args]
+    end
+    ruby2_keywords(:initialize)
```

Comment:
What's the purpose of this code above? It seems like it's trying to handle many different cases, but why?

Reply:
A core change here is to make the Job instance a meaningful entity to Sidekiq. Currently, it serves little to no purpose. A Sidekiq job has 2 key bits of information it needs to be executed: its options (which queue, how many retries, etc.) and its arguments. Both are optional. You can have a Sidekiq job in 4 flavors:

- no opts, no args
- opts, no args
- no opts, args
- opts, args

Moreover, while in the end all arguments need to be in an Array, many Jobs only have 1 argument. So, for every flavor where there is an argument, there are 2 additional flavors:

- arg
- [arg_1, arg_2]

I want to create an `initialize` method signature that works for all 6 flavors:

- `Job.new()`
- `Job.new(queue: 'some_queue')`
- `Job.new(record.id)`
- `Job.new(record.id, 'another_arg')`
- `Job.new({queue: 'some_queue'}, record.id)`
- `Job.new({queue: 'some_queue'}, record.id, 'another_arg')`

This is my current best attempt at an `initialize` method that does what I want in each of those 6 scenarios.

Reply:
Do you ever come back to a problem and wonder what exact thought-path led you to the result you came to the first time around? I have no idea why I got it in my head to put the options hash first and the arguments array second in the method signature, but that is what made this method so weird. Coming back to it, I realized that I should simply reverse that order and use normal Ruby. Sorry for creating some consternation at the outset, should've always looked like what it is now.